### PR TITLE
Add QNN Op package #1

### PR DESCRIPTION
**Pull request by jijoongmoon**
This PR adds the source code and directories for qualcomm npu op
support.
 . nntrainer / npu / qnn

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/qnn_context.cpp [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  str_map[assigned_key] = factory;
+  int_map[assigned_int_key] = assigned_key;
+
+  ml_logd("cl_context: factory has registered with key: %s, int_key: %d",
+          assigned_key.c_str(), assigned_int_key);
+
+  return assigned_int_key;
+}
+
+void ClContext::initBlasClKernels() {
```
Comment (skykongkong8) [inline nntrainer/qnn_context.cpp:140]: Is this a temporary change due to excessive amount of code in all commits for QNN(including later PRs)?

Comment (jijoongmoon) [reply nntrainer/qnn_context.cpp:140]: This will be chaged in later PR.

File: meson_options.txt [hunk 2]
```diff
@@ -45,6 +45,7 @@ option('enable-avx', type: 'boolean', value: true)
 option('enable-opencl', type: 'boolean', value: false)
 option('enable-biqgemm', type: 'boolean', value: false)
 option('enable-benchmarks', type: 'boolean', value : false)
+option('enable-qnn', type: 'boolean', value: true)
```
Comment (EunjuYang) [inline meson_options.txt:48]: What about setting the default value of `enable-qnn` as false?

File: nntrainer/qnn_context.h [hunk 3]
```diff
@@ -0,0 +1,299 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2024 Debadri Samaddar <s.debadri@samsung.com>
+ *
+ * @file    cl_context.h
+ * @date    23 Feb 2024
+ * @see     https://github.com/nnstreamer/nntrainer
+ * @author  Debadri Samaddar <s.debadri@samsung.com>
```
Comment (EunjuYang) [inline nntrainer/qnn_context.h:8]: It seems that you need to update the Copyright for this qnn_context.h

File: nntrainer/qnn_context.h [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+#include <opencl_command_queue_manager.h>
+#include <opencl_context_manager.h>
+#include <opencl_kernel.h>
+#include <opencl_program.h>
+
+#include <nntrainer_log.h>
+
+namespace nntrainer {
+
+extern std::mutex cl_factory_mutex;
```
Comment (EunjuYang) [inline nntrainer/qnn_context.h:41]:
```suggestion
extern std::mutex qnn_factory_mutex;
```

File: nntrainer/qnn_context.cpp [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+#include <concat_cl.h>
+#include <fc_layer_cl.h>
+#include <reshape_cl.h>
+#include <rmsnorm_layer_cl.h>
+#include <swiglu_cl.h>
+#include <transpose_cl.h>
+
+namespace nntrainer {
+
+std::mutex cl_factory_mutex;
```
Comment (EunjuYang) [inline nntrainer/qnn_context.cpp:30]: ditto.
```suggestion
std::mutex qnn_factory_mutex;
```

File: nntrainer/qnn_context.h [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+ * @see     https://github.com/nnstreamer/nntrainer
+ * @author  Debadri Samaddar <s.debadri@samsung.com>
+ * @bug     No known bugs except for NYI items
+ * @brief   This file contains app context related functions and classes that
+ * manages the global configuration of the current OpenCL environment. It also
+ * creates the OpenCL command queue and context.
+ */
+
+#ifndef __CL_CONTEXT_H__
+#define __CL_CONTEXT_H__
```
Comment (EunjuYang) [inline nntrainer/qnn_context.h:16]: It should be updated for `qnn_context.h`.
```suggestion
#ifndef __QNN_CONTEXT_H__
#define __QNN_CONTEXT_H__
```

File: nntrainer/qnn_context.cpp (hunk 1 above)
Comment (EunjuYang) [reply nntrainer/qnn_context.cpp:140]: I just check this comment.
I will skip to review the `qnn_context` for this PR.

**Pull request closed without merging**